README
Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero is
Copyright (C) 2022 by Thomas R. Wyant, III

DESCRIPTION

This package contains a Perl::Critic policy to prohibit the use of numeric
variable names that start with 0, except for $0 itself.

Perl::Critic is a prerequisite for this package. It is the intent that
the prerequisites of this package be those of Perl::Critic, so no extra
stuff should be needed.

INSTALLATION

This package is installable by either of the two usual incantations:

    tar -xzf \
	Perl-Critic-Policy-Variables-ProhibitUnusedVarsStricter-9.999.tar
    perl Makefile.PL
    make
    make test
    make install

or

    tar -xzf \
	Perl-Critic-Policy-Variables-ProhibitUnusedVarsStricter-9.999.tar
    perl Build.PL
    ./Build
    ./Build test
    sudo ./Build install

You should substitute the appropriate program name for 'make', eg nmake
(typically) under MSWin32, or mms or mmk under VMS. See
ftp://ftp.microsoft.com/Softlib/MSLFILES/nmake15.exe for a copy of
nmake for MSWin32 if you need it.

Of course, since it's pure Perl, you can just expand the kit and drop
the .pm file into the Perl/Critic/Policy/Variables/ directory (creating
it if necessary) in the appropriate place in your @INC directories.

LICENSING INFORMATION

This package is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses in the directory LICENSES.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.


Changes
0.001		2022-07-25	T. R. Wyant
    Initial release to CPAN.


Build.PL
use strict;
use warnings;

use 5.006001;

use lib 'inc';

use Module::Build;
use My::Module::Build;
use My::Module::Meta qw{
    build_required_module_versions
    meta_merge
    recommended_module_versions
    required_module_versions
    requires_perl
};

( my $mbv = Module::Build->VERSION() ) =~ s/_//g;

my %args = (
    module_name         => 'Perl::Critic::Policy::Variables::ProhibitNumericNamesWithLeadingZero',
    dist_author         => 'Thomas R. Wyant, III (wyant at cpan dot org)',
    # dist_abstract       => 'Critique unused variables in Perl source',
    license             => 'perl',
    create_packlist     => 1,
    sign                => 0,

    requires            => { required_module_versions(
                                perl => requires_perl(),
                            ) },
    build_requires      => { build_required_module_versions() },
    recommends          => { recommended_module_versions() },

    add_to_cleanup      => [
        qw<
            Debian_CPANTS.txt
        >,
    ],
);

$mbv >= 0.28
    and $args{meta_merge} = meta_merge();

$mbv >= 0.34
    and $args{auto_configure_requires} = 0;	# Don't require Module::Build

my $builder = My::Module::Build->new(
    %args,
);

$builder->create_build_script();

##############################################################################
# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=78 ft=perl expandtab shiftround :


Makefile.PL
use strict;
use warnings;

use 5.006001;

use lib qw{ inc };

use ExtUtils::MakeMaker;
use My::Module::Meta qw{
    build_required_module_versions
    meta_merge
    required_module_versions
    requires_perl
    recommended_module_versions
};

( my $mmv = ExtUtils::MakeMaker->VERSION ) =~ s/_//g;

my %args = (
    NAME            => 'Perl::Critic::Policy::Variables::ProhibitNumericNamesWithLeadingZero',
    VERSION_FROM    => 'lib/Perl/Critic/Policy/Variables/ProhibitNumericNamesWithLeadingZero.pm',
    ABSTRACT_FROM    => 'lib/Perl/Critic/Policy/Variables/ProhibitNumericNamesWithLeadingZero.pm',
    PREREQ_PM       => { required_module_versions() },
    PL_FILES        => {},        # Prevent old MakeMaker from running Build.PL
    AUTHOR          => 'Tom Wyant (wyant at cpan dot org)',
);

$mmv >= 6.31
    and $args{LICENSE} = 'perl';

$mmv >= 6.4501
    and $args{META_MERGE} = meta_merge();

$mmv >= 6.4701
    and $args{MIN_PERL_VERSION} = requires_perl();

$mmv >= 6.5501
    and $args{BUILD_REQUIRES} = { build_required_module_versions() }
    or $mmv >= 6.4501
    and $args{META_MERGE}{build_requires} = { build_required_module_versions() };

WriteMakefile( %args );

sub MY::postamble {
    # my ( $self, @args ) = @_;
    my ( $self ) = @_;

    my $authortest = $self->test_via_harness(
	'$(FULLPERLRUN)', '$(AUTHORTEST_FILES)' );
    $authortest =~ s/ \s+ \z //smx;
    $authortest =~ s/ \A \s+ //smx;
    chomp $authortest;

    return <<"EOD";

AUTHORTEST_FILES = t/*.t xt/author/*.t

authortest :: pure_all
	AUTHOR_TESTING=1 $authortest

testcover :: pure_all
	cover -test
EOD
}

##############################################################################
# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=78 ft=perl expandtab shiftround :


inc/My/Module/Build.pm
package My::Module::Build;

use 5.006001;

use strict;
use warnings;

use Carp;

use Module::Build;
our @ISA = qw{ Module::Build };

our $VERSION = '0.001';

sub ACTION_authortest {
    my ( $self, @args ) = @_;

    local $ENV{AUTHOR_TESTING} = 1;

    $self->depends_on( 'build' );
    $self->test_files( qw{ t xt/author } );
    $self->SUPER::ACTION_test( @args );

    return;

}

sub ACTION_genclean {
    my ( $self ) = @_;

    $self->depends_on( 'realclean' );

    foreach my $fn ( qw{ META.json META.yml } ) {
        -f $fn
            and unlink $fn;
    }

    return;
}


1;

__END__

=head1 NAME

My::Module::Build - Customize L<Module::Build|Module::Build>.

=head1 SYNOPSIS

    use lib qw{ inc };
    use My::Module::Build;
    my $mb = My::Module::Build->new(
	module_name	=> 'Foo::Bar',
	dist_author	=> 'Yehudi Menuhin',
	dist_abstract	=> 'Foo some bars',
	license		=> 'perl',
    );
    $mb->create_build_script();

=head1 DESCRIPTION

This subclass of L<Module::Build|Module::Build> adds build targets as
convenient.

=head1 ACTIONS

=over

=item authortest

This action performs both the normal tests in F<t/> and the author tests
in F<xt/author/>.

=item genclean

This action is just like the C<realclean> action, but also removes files
generated by the packaging process. These are F<META.json> and F<META.yml>.

=back

=head1 SEE ALSO

L<Module::Build|Module::Build>.

=head1 SUPPORT

Support is by the author. Please file bug reports at
L<https://rt.cpan.org>, or in electronic mail to the author.

=head1 AUTHOR

Thomas R. Wyant, III F<wyant at cpan dot org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2022 by Thomas R. Wyant, III

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses in the directory LICENSES.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut

##############################################################################
# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=78 ft=perl expandtab shiftround :


lib/Perl/Critic/Policy/Variables/ProhibitNumericNamesWithLeadingZero.pm
package Perl::Critic::Policy::Variables::ProhibitNumericNamesWithLeadingZero;

use 5.006001;
use strict;
use warnings;

use English qw{ -no_match_vars };

use PPIx::QuoteLike 0.011;
use PPIx::QuoteLike::Constant 0.011 qw{
    LOCATION_LINE
    LOCATION_LOGICAL_LINE
    LOCATION_CHARACTER
};

use PPIx::QuoteLike 0.011;
use PPIx::Regexp 0.071;
use Readonly;
# use Scalar::Util qw{ refaddr };

use Perl::Critic::Utils qw< :booleans :characters hashify :severities >;

use base 'Perl::Critic::Policy';

our $VERSION = '0.001';

#-----------------------------------------------------------------------------

Readonly::Scalar my $DESC => q<Numeric variable name %s starts with 0>;
Readonly::Scalar my $EXPL =>
    q<Numeric variable names may not start with 0, except for $0 itself>; ## no critic (RequireInterpolationOfMetachars)

Readonly::Scalar my $PACKAGE    => '_' . __PACKAGE__;

Readonly::Scalar my $LEFT_BRACE => q<{>;    # } Seems P::C::U should have

Readonly::Hash my %IS_COMMA     => hashify( $COMMA, $FATCOMMA );
Readonly::Hash my %LOW_PRECEDENCE_BOOLEAN => hashify( qw{ and or xor } );

Readonly::Array my @DOUBLE_QUOTISH => qw{
    PPI::Token::Quote::Double
    PPI::Token::Quote::Interpolate
    PPI::Token::QuoteLike::Backtick
    PPI::Token::QuoteLike::Command
    PPI::Token::QuoteLike::Readline
    PPI::Token::HereDoc
};
Readonly::Array my @REGEXP_ISH => qw{
    PPI::Token::Regexp::Match
    PPI::Token::Regexp::Substitute
    PPI::Token::QuoteLike::Regexp
};

#-----------------------------------------------------------------------------

sub supported_parameters { return }

sub default_severity     { return $SEVERITY_MEDIUM       }
sub default_themes       { return qw< trw bug maintenance >  }
sub applies_to           { return qw< PPI::Document >    }

#-----------------------------------------------------------------------------

sub violates {
#   my ( $self, $elem, $document ) = @_;
    my ( $self, undef, $document ) = @_;

    return ( map { $_->[0] }
        sort { $a->[1] <=> $b->[1] || $a->[2] <=> $b->[2] }
        map { [ $_, $_->line_number(), $_->column_number() ] }
        $self->_critique_element( $document )
    );
}

#-----------------------------------------------------------------------------

sub _critique_element {
    my ( $self, $elem ) = @_;

    my @violations;

    # Find $0nnn
    push @violations, $self->_find_bare_violations( $elem );

    # Find ${0nnn}
    push @violations, $self->_find_bracketed_violations( $elem );

    # Find "$0nnn" and "${0nnn}"
    push @violations, $self->_find_string_violations( $elem );

    # Find m/$0nnn/ and m/${0nnn}/
    push @violations, $self->_find_regex_violations( $elem );

    return @violations;
}

#-----------------------------------------------------------------------------

# Find $0nnn
sub _find_bare_violations {
    my ( $self, $elem ) = @_;

    my @violations;

    # Find $0nnn
    foreach my $sym ( @{ $elem->find( 'PPI::Token::Magic' ) || [] } ) {
        my $name = $sym->symbol();
        $name =~ m/ \W 0 [0-9]+ \z /smx ## no critic (ProhibitEnumeratedClasses)
            or next;
        push @violations, $self->violation(
            sprintf( $DESC, $name ),
            $EXPL,
            $sym,
        );
    }

    return @violations;
}

#-----------------------------------------------------------------------------

# Find ${0nnn}
sub _find_bracketed_violations {
    my ( $self, $elem ) = @_;

    my @violations;

    foreach my $cast ( @{ $elem->find( 'PPI::Token::Cast' ) || [] }) {
        my $block = $cast->snext_sibling()
            or next;
        $block->isa( 'PPI::Structure::Block' )
            or next;
        my @tokens = @{ $block->find( 'PPI::Token' ) || [] }
            or next;
        @tokens > 1
            and next;
        $tokens[0]->isa( 'PPI::Token::Number' )
            or next;
        $tokens[0]->content() =~ m/ \A 0 [0-9]+ \z /smx ## no critic (ProhibitEnumeratedClasses)
            or next;
        push @violations, $self->violation(
            sprintf( $DESC, join q<>, $cast->content(), $block->content() ),
            $EXPL,
            $cast,
        );
    }

    return @violations;
}

#-----------------------------------------------------------------------------

# Find "$0nnn" and "${0nnn}"
sub _find_string_violations {
    my ( $self, $elem ) = @_;

    my @violations;

    foreach my $class ( @DOUBLE_QUOTISH ) {
        foreach my $ppi_str ( @{ $elem->find( $class ) || [] } ) {
            my $ppix_str = PPIx::QuoteLike->new( $ppi_str )
                or next;
            foreach my $interp (
                @{ $ppix_str->find( 'PPIx::QuoteLike::Token::Interpolation' ) || [] }
            ) {
                # NOTE that policy Variables::ProhibitUnusedVarsStricter
                # uses a wrapper for $elem->ppi() because it has to link
                # the little PPI documents it makes out of strings to
                # the parent PPI document. This is enforced by test
                # xt/author/require_wrapper.t. This policy has (so far)
                # no need to link the two documents together. If it
                # develops the need, copy this test in, fix all
                # violatioms, and be prepared to rewrite calls to
                # parent() and top().
                push @violations, $self->_critique_element( $interp->ppi() );
            }
        }
    }

    return @violations;
}

#-----------------------------------------------------------------------------

# Find m/$0nnn/ and m/${0nnn}/
sub _find_regex_violations {
    my ( $self, $elem ) = @_;

    my @violations;

    foreach my $class ( @REGEXP_ISH ) {
        foreach my $ppi_re ( @{ $elem->find( $class ) || [] } ) {
            my $ppix_re = PPIx::Regexp->new( $ppi_re )
                or next;
            foreach my $code (
                @{ $ppix_re->find( 'PPIx::Regexp::Token::Code' ) || [] }
            ) {
                # NOTE see previous note.
                push @violations, $self->_critique_element( $code->ppi() );
            }
        }
    }

    return @violations;
}

#-----------------------------------------------------------------------------

1;

__END__

#-----------------------------------------------------------------------------

=pod

=head1 NAME

Perl::Critic::Policy::Variables::ProhibitNumericNamesWithLeadingZero - Don't use numeric variable names with leading zeroes.


=head1 AFFILIATION

This Policy is stand-alone, and is not part of the core
L<Perl::Critic|Perl::Critic>.


=head1 DESCRIPTION

Numeric variable names with leading zeroes are unsupported by Perl, and
can lead to obscure bugs. In particular, they are not (or not
straightforwardly) accessible as C<${0nnn}>.

Starting with Perl 5.32, these variables represent a syntax error, so
this policy is useless with current Perls. On the other hand, it may be
useful for those with an older code base, especially if they are
preparing to upgrade it.

=head1 CONFIGURATION

This policy supports no configuration items above and beyond the
standard ones.

=head1 SUPPORT

Support is by the author. Please file bug reports at
L<https://rt.cpan.org/Public/Dist/Display.html?Name=Perl-Critic-Policy-Variables-ProhibitUnusedVarsStricter>,
L<https://github.com/trwyant/perl-Perl-Critic-Policy-Variables-ProhibitUnusedVarsStricter/issues>, or in
electronic mail to the author.

=head1 AUTHOR

Thomas R. Wyant, III F<wyant at cpan dot org>

=head1 COPYRIGHT

Copyright (C) 2022 Thomas R. Wyant, III

=head1 LICENSE

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses in the directory LICENSES.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 72
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=72 ft=perl expandtab shiftround :


inc/My/Module/Meta.pm
package My::Module::Meta;

use 5.006001;

use strict;
use warnings;

use Carp;

our $VERSION = '0.001';

sub new {
    my ( $class ) = @_;
    ref $class and $class = ref $class;
    my $self = {
	distribution => $ENV{MAKING_MODULE_DISTRIBUTION},
    };
    bless $self, $class;
    return $self;
}

sub abstract {
    return q<Don't use numeric variable names with leading zeroes.>;
}

sub add_to_cleanup {
    return [ qw{ cover_db } ];
}

sub author {
    return 'Thomas R. Wyant, III (wyant at cpan dot org)';
}

sub build_requires {
    return +{
        'Carp'      => 0,
	'Perl::Critic::TestUtils'	=> 0,
	'PPI::Document'			=> 0,
	'Test::More'	=> 0.88,	# for done_testing();
        'Test::Perl::Critic::Policy'    => 0,
        lib         => 0,
    };
}

sub configure_requires {
    return +{
	'lib'	=> 0,
	'strict'	=> 0,
	'warnings'	=> 0,
    };
}

sub dist_name {
    return 'Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero';
}


sub license {
    return 'perl';
}

sub meta_merge {
    my ( undef, @extra ) = @_;
    return {
	'meta-spec'	=> {
	    version	=> 2,
	},
	dynamic_config	=> 1,
	resources	=> {
	    bugtracker	=> {
		web	=> 'https://rt.cpan.org/Public/Dist/Display.html?Name=Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero',
		# web	=> 'https://github.com/trwyant/perl-Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero/issues',
		mailto  => 'wyant@cpan.org',
	    },
	    license	=> 'http://dev.perl.org/licenses/',
	    repository	=> {
		type	=> 'git',
		url	=> 'git://github.com/trwyant/perl-Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero.git',
		web	=> 'https://github.com/trwyant/perl-Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero',
	    },
	},
	@extra,
    };
}


sub module_name {
    return 'Perl::Critic::Policy::Variables::ProhibitNumericNamesWithLeadingZero';
}

sub no_index {
    return +{
	directory => [ qw{ doc examples inc tools xt } ],
	file	=> [ qw{ TODO.pod } ],
    };
}

sub provides {
    my $provides;
    local $@ = undef;

    eval {
	require CPAN::Meta;
	require ExtUtils::Manifest;
	require Module::Metadata;

	my $manifest;
	{
	    local $SIG{__WARN__} = sub {};
	    $manifest = ExtUtils::Manifest::maniread();
	}
	keys %{ $manifest || {} }
	    or return;

	# Skeleton so we can use should_index_file() and
	# should_index_package().
	my $meta = CPAN::Meta->new( {
		name	=> 'Euler',
		version	=> 2.71828,
		no_index	=> no_index(),
	    },
	);

	# The Module::Metadata docs say not to use
	# package_versions_from_directory() directly, but the 'files =>'
	# version of provides() is broken, and has been known to be so
	# since 2014, so it's not getting fixed any time soon. So:

	foreach my $fn ( sort keys %{ $manifest } ) {
	    $fn =~ m/ [.] pm \z /smx
		or next;
	    my $pvd = Module::Metadata->package_versions_from_directory(
		undef, [ $fn ] );
	    foreach my $pkg ( keys %{ $pvd } ) {
		$meta->should_index_package( $pkg )
		    and $meta->should_index_file( $pvd->{$pkg}{file} )
		    and $provides->{$pkg} = $pvd->{$pkg};
	    }
	}

	1;
    } or return;

    return ( provides => $provides );
}

sub requires {
    my ( undef, @args ) = @_;
    return {
	'English'		    => 0,
	# 'Perl::Critic::Document'    => 1.119,   # need 1.119 here
        'Perl::Critic::Policy'      => 1.119,
        'Perl::Critic::Utils'       => 1.119,
	'PPI::Document'		    => 0,
	# 'PPI::Token::Symbol'        => 0,
        'PPIx::QuoteLike'           => 0.011,   # For full scope inside ""
	'PPIx::QuoteLike::Constant' => 0.011,
        'PPIx::Regexp'              => 0.071,   # For full scope inside //
        'Readonly'                  => 0,
        base                        => 0,
        strict                      => 0,
        warnings                    => 0,
        @args,
    };
}

sub recommended_module_versions {
    return (
	# 'File::Which'   => 0,
    );
}

sub requires_perl {
    return '5.006001';
}

sub script_files {
    return [
    ];
}

sub version_from {
    return 'lib/Perl/Critic/Policy/Variables/ProhibitNumericNamesWithLeadingZero.pm';
}

1;

__END__

=head1 NAME

My::Module::Meta - Metadata for the current module

=head1 SYNOPSIS

 use lib qw{ inc };
 use My::Module::Meta qw{ recommended_module_versions };

=head1 DESCRIPTION

This Perl module holds metadata for the current module. It is private to
the current module.

=head1 METHODS

This class supports the following public methods:

=head2 new

 my $meta = My::Module::Meta->new();

This method instantiates the class.

=head2 build_requires

 use YAML;
 print Dump( $meta->build_requires() );

This method computes and returns a reference to a hash describing the
modules required to build the C<App::Retab> package, suitable for
use in a F<Build.PL> C<build_requires> key, or a F<Makefile.PL>
C<< {META_MERGE}->{build_requires} >> or C<BUILD_REQUIRES> key.

=head2 meta_merge

 use YAML;
 print Dump( $meta->meta_merge() );

This method returns a reference to a hash describing the meta-data which
has to be provided by making use of the builder's C<meta_merge>
functionality. This includes the C<dynamic_config>, C<no_index> and
C<resources> data.

=head2 recommended_module_versions

This subroutine returns an array of the names and versions of
recommended modules.

=head2 requires

 use YAML;
 print Dump( $meta->requires() );

This method computes and returns a reference to a hash describing
the modules required to run the C<App::Retab>
package, suitable for use in a F<Build.PL> C<requires> key, or a
F<Makefile.PL> C<PREREQ_PM> key. Any additional arguments will be
appended to the generated hash. In addition, unless
L<distribution()|/distribution> is true, configuration-specific modules
may be added.

=head2 requires_perl

This subroutine returns the version of Perl required by the module.

=head1 SUPPORT

Support is by the author. Please file bug reports at
L<https://rt.cpan.org/Public/Dist/Display.html?Name=Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero>,
L<https://github.com/trwyant/perl-Perl-Critic-Policy-Variables-ProhibitNumericNamesWithLeadingZero/issues>, or in
electronic mail to the author.

=head1 AUTHOR

Thomas R. Wyant, III F<wyant at cpan dot org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013-2022 by Thomas R. Wyant, III

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the full text
of the licenses in the directory LICENSES.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut

##############################################################################
# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 72
#   indent-tabs-mode: nil
#   c-indentation-style: bsd
# End:
# ex: set ts=8 sts=4 sw=4 tw=72 ft=perl expandtab shiftround :

=head2 abstract

This method returns the distribution's abstract.

=head2 add_to_cleanup

This method returns a reference to an array of files to be added to the
cleanup.

=head2 author

This method returns the name of the distribution author

=head2 configure_requires

 use YAML;
 print Dump( $meta->configure_requires() );

This method returns a reference to a hash describing the modules
required to configure the package, suitable for use in a F<Build.PL>
C<configure_requires> key, or a F<Makefile.PL>
C<< {META_MERGE}->{configure_requires} >> or C<CONFIGURE_REQUIRES> key.

=head2 dist_name

This method returns the distribution name.

=head2 license

This method returns the distribution's license.

=head2 meta_merge

 use YAML;
 print Dump( $meta->meta_merge() );

This method returns a reference to a hash describing the meta-data which
has to be provided by making use of the builder's C<meta_merge>
functionality. This includes the C<dynamic_config> and C<resources>
data.

Any arguments will be appended to the generated array.

=head2 module_name

This method returns the name of the module the distribution is based
on.

=head2 no_index

This method returns the names of things which are not to be indexed
by CPAN.

=head2 provides

 use YAML;
 print Dump( [ $meta->provides() ] );

This method attempts to load L<Module::Metadata|Module::Metadata>. If
this succeeds, it returns a C<provides> entry suitable for inclusion in
L<meta_merge()|/meta_merge> data (i.e. C<'provides'> followed by a hash
reference). If it can not load the required module, it returns nothing.

=head2 script_files

This method returns a reference to an array containing the names of
script files provided by this distribution. This array may be empty.

=head2 version_from

This method returns the name of the distribution file from which the
distribution's version is to be derived.

